core/src/main/java/org/pentaho/di/connections/vfs/provider/ConnectionFileNameParser.java

```diff
@@ -22,36 +22,112 @@
 
 package org.pentaho.di.connections.vfs.provider;
 
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
 import org.apache.commons.vfs2.FileName;
 import org.apache.commons.vfs2.FileSystemException;
 import org.apache.commons.vfs2.FileType;
-import org.apache.commons.vfs2.provider.AbstractFileName;
 import org.apache.commons.vfs2.provider.AbstractFileNameParser;
-import org.apache.commons.vfs2.provider.FileNameParser;
 import org.apache.commons.vfs2.provider.UriParser;
 import org.apache.commons.vfs2.provider.VfsComponentContext;
 
+import static org.pentaho.di.connections.vfs.provider.ConnectionFileObject.DELIMITER_CHAR;
+
 public class ConnectionFileNameParser extends AbstractFileNameParser {
-  @Override public FileName parseUri( VfsComponentContext vfsComponentContext, FileName fileName, String uri )
+  private static final String[] SCHEMES = new String[] { ConnectionFileProvider.SCHEME };
+
+  /**
+   * Parses a PVFS URI.
+   * <h3>Structure of a PVFS URI</h3>
+   * The general structure of a PVFS URI is: <code>pvfs://[(connection-name)[/(path...)]/?]</code>.
+   * The path separators after the connection name tolerate the "\" character as equivalent to the "/" character.
+   * <h3>Connection names</h3>
+   * <p>
+   * Connection names cannot contain the characters "\" and "/".
+   * <p>
+   * In a PVFS URI, connection names occupy the authority/host segment. Given the character set of connection names,
+   * which includes, for example, spaces and international characters, a PVFS URI cannot be represented as a standard
+   * URL or URI, neither be directly parsed by a standard URL or URI parser.
+   * <p>
+   * Standard URLs or URIs do not admit percent-encoding in the host segment. And, despite URIs supporting the
+   * <a href="https://datatracker.ietf.org/doc/html/rfc3490">Internationalized Domain Names (IDNA)</a> encoding in the
+   * host segment, providing support for a broad set of non-ASCII international characters, something as simple as a
+   * space character is still not supported. It is in part due to this characteristic of URLs and URIs that certain
+   * standard protocols, such as `file`, have an empty authority segment, `file:///`.
+   *
+   * @param vfsComponentContext The VFS component context.
+   * @param baseFileName        The base file name.
+   * @param pvfsUri             The PVFS URI to parse.
+   * @return The connection file name.
+   */
+  @Override
+  public FileName parseUri( VfsComponentContext vfsComponentContext, FileName baseFileName, String pvfsUri )
     throws FileSystemException {
-    return parseUri( uri, this );
+    return parseUri( pvfsUri );
   }
 
-  public static AbstractFileName parseUri( String uri, FileNameParser fileNameParser ) throws FileSystemException {
+  @NonNull
+  public ConnectionFileName parseUri( String pvfsUri ) throws FileSystemException {
```

Comment:
I mostly just added comments with examples.
There was one actual change: switching the order by which `UriParser#extractFirstElement(.)` and `UriParser#normalisePath()` are called.
The reason is so that relative paths trying to go above the connection are not successful: `pvfs://connection/../other-connection/folder`.